cmd/otel-allocator/collector/collector.go

```diff
@@ -87,6 +86,9 @@ func (k *Client) Watch(ctx context.Context, labelMap map[string]string, fn func(
 
 	go func() {
 		for {
+			// add timeout to the context before calling Watch
+			ctx, cancel := context.WithTimeout(ctx, watcherTimeout)
+			defer cancel()
```

Comment:
Is this going to accumulate deferred executions for every iteration through this infinite loop?  Do we need to defer execution here?  Can it simply be not executed or can it be executed without deferral at the end of each loop iteration?

Reply:
Thanks for the review! Hmm I think if the operation returns before the timeout (no timeout triggered) then a call to cancel() is needed for cleanup.  I agree I shouldn't pile up a bunch of defer statements and should call `cancel()` as soon as possible.  If I add cancel() to the end of for loop and before the returns, I think there's still a chance if a panic occurs `cancel()` won't be called without `defer`.

To solve this I wrapped this whole block into its own method called `restartWatch` and that way `defer` can be called properly. Let me know if you have any thoughts on this!